legacy-sdk/whirlpool/tests/utils/litesvm.ts

```diff
@@ -0,0 +1,1230 @@
+    }
+    // Check for other Token-2022 extension errors
+    if (log.includes("Extension") && log.includes("Error")) {
+      return log;
+    }
+    // Extract anchor errors
+    if (log.includes("AnchorError") || log.includes("Error Code:")) {
+      return log;
+    }
+  }
+  return null;
+}
+
+/**
+ * Map LiteSVM errors to more meaningful error messages
+ * Token-2022 and other programs may return undefined or poorly serialized errors in LiteSVM
+ */
+function mapLiteSVMError(error: any, logs: string[]): string {
+  let errorStr = error.toString();
+  // Map common Solana error codes to human-readable messages
+  if (errorStr === "6") {
+    errorStr = "6 (NotEnoughAccountKeys)";
+  } else if (errorStr === "7") {
+    errorStr = "7 (AccountBorrowFailed)";
+  } else if (errorStr === "11") {
+    errorStr = "11 (signature verification fail)";
+  }
+  // If error serialization failed, try to extract from logs or error structure
+  if (errorStr === "undefined" || !errorStr || errorStr === "[object Object]") {
+    // Try to extract from logs first
+    const errorFromLogs = extractErrorFromLogs(logs);
+    if (errorFromLogs) {
+      return errorFromLogs;
+    }
+    // Try to extract from error object structure
+    if (error && typeof error === "object" && error.InstructionError) {
+      const [index, innerError] = error.InstructionError;
+      if (innerError?.Custom !== undefined) {
+        return `InstructionError(${index}, Custom(${innerError.Custom}))`;
+      }
+    }
+    // Empty logs + undefined error typically means Token-2022 NonTransferable blocked a transfer
+    if (logs.length === 0) {
+      return "Transfer is disabled for this mint";
+    }
+  }
+  return errorStr;
+}
+
+/**
+ * Reset the LiteSVM instance, forcing a fresh blockchain state
+ * Use this in beforeEach() hooks to get isolated state between tests
+ */
+export async function resetLiteSVM(): Promise<void> {
+  _litesvm = null;
+  _logListeners = [];
+  _nextListenerId = 1;
+  await startLiteSVM();
+}
+
+/**
+ * Create a LiteSVM-powered Anchor provider
+ */
+export async function createLiteSVMProvider(): Promise<anchor.AnchorProvider> {
+  const litesvm = await startLiteSVM();
+  // Create wallet
+  const wallet = Keypair.generate();
+  // Fund the wallet using airdrop (500 SOL for tests that need large transfers)
+  litesvm.airdrop(wallet.publicKey, BigInt(500e9));
+  // Create connection wrapper
+  const connection = createLiteSVMConnection(litesvm);
+  // Create provider
+  // Connection object implements the subset of Connection methods needed by Anchor
+  return new anchor.AnchorProvider(
+    // @ts-expect-error - Connection interface is partially implemented
+    connection,
+    new anchor.Wallet(wallet),
+    { commitment: "confirmed" },
+  );
+}
+
+/**
+ * Connection wrapper for LiteSVM
+ * Uses getLiteSVM() dynamically to support instance resets
+ */
+function createLiteSVMConnection(litesvm: LiteSVM) {
```

Comment:
🤯 nice work here. there is a lot of code here. seems like a great addition to a utility crate/package in the main repo - not saying there's any work to do here, just interesting that it's not part of the core repo